akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala

```diff
@@ -924,10 +847,30 @@ private[remote] abstract class ArteryTransport(_system: ExtendedActorSystem, _pr
     }
   }
 
+  // Checks for termination hint messages and sends an ACK for those (not processing them further)
+  // Purpose of this stage is flushing, the sender can wait for the ACKs up to try flushing
+  // pending messages.
+  val flushReplier: Flow[InboundEnvelope, InboundEnvelope, NotUsed] = {
+    Flow[InboundEnvelope].filter { envelope =>
+      envelope.message match {
+        case Flush =>
+          envelope.sender match {
+            case OptionVal.Some(snd) =>
+              snd.tell(FlushAck, ActorRef.noSender)
+            case OptionVal.None =>
+              log.error("Expected sender for Flush message")
+          }
+          false
+        case _ => true
+      }
+    }
+  }
```

Comment:
So this runs inside the regular and control inbound, acking and going through the outgoing regular with the ack, that still doesn't guarantee the Terminated will come after all messages though, does it, just makes it much much less likely?

Reply:
I don't think we have any async boundaries after this so when reply is sent from one lane the preceding message has already been delivered to the mailbox of the destination actor. Is that thinking correct?

Reply:
Not sure that is a guarantee, but extremely likely, then adding the network hop of the ack on top the message should have ended up in the inbox...